File: Tree/C/M513-find-bottom-left-tree-value.c
// time:  O(n)
// space: O(h)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void bottomLeft(struct TreeNode* root, int* maxDepth, int* res, int depth){
    if(!root)
        return;
    
    if(depth > *maxDepth){
        *res = root->val;
        *maxDepth = depth;
    }
    
    bottomLeft(root->left, maxDepth, res, depth+1);
    bottomLeft(root->right, maxDepth, res, depth+1);
}
int findBottomLeftValue(struct TreeNode* root) {
    int* maxDepth = (int*)malloc(sizeof(int));
    *maxDepth = 0;
    int* res = (int*)malloc(sizeof(int));
    *res = 0;
    int depth = 1;
    
    bottomLeft(root, maxDepth, res, depth);
    return *res;
}

File: Tree/Python3/M513-find-bottom-left-tree-value.py
# time:  O(n)
# space: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        nl = [root]
        while(1):
            tnl = []
            
            for n in nl:
                if n.left:
                    tnl.append(n.left)
                if n.right:
                    tnl.append(n.right)
                    
            if tnl:
                nl = tnl
            else:
                return nl[0].val